aten/src/ATen/Context.h

```diff
@@ -424,6 +426,13 @@ static inline void manual_seed(uint64_t seed) {
       }
     }
   }
+
+  if (hasMPS()) {
+    auto mps_gen = globalContext().defaultGenerator(DeviceType::MPS);
+    // See Note [Acquire lock when using random generators]
+    std::lock_guard<std::mutex> lock(mps_gen.mutex());
```

Comment:
Why is this lock outside of the object? Why isn't that handled within mps_gen?

Reply:
The lock depends on a Mutex within mps_gen itself. The note `[Acquire lock when using random generators]` explains why it's designed this way globally. It's used to protect setting the seeds and other random params. 
CUDA is using the same logic in the lines above this code:
```
auto cuda_gen = globalContext().defaultGenerator(
          Device(at::kCUDA, static_cast<c10::DeviceIndex>(i)));
      {
        // See Note [Acquire lock when using random generators]
        std::lock_guard<std::mutex> lock(cuda_gen.mutex());
        cuda_gen.set_current_seed(seed);
      }  
```